android-components/components/feature/media/src/main/java/mozilla/components/feature/media/fullscreen/MediaSessionFullscreenFeature.kt

```diff
@@ -35,58 +35,64 @@ class MediaSessionFullscreenFeature(
             flow.map {
                 it.tabs + it.customTabs
             }.map { tab ->
-                tab.filter { it.mediaSessionState != null && it.mediaSessionState!!.fullscreen }
-            }.distinctUntilChanged().collect { states ->
-                processFullscreen(states)
-                processDeviceSleepMode(states)
+                tab.firstOrNull { it.mediaSessionState?.fullscreen == true }
+            }.distinctUntilChanged { old, new ->
+                old.hasSameOrientationInformationAs(new)
+            }.collect { state ->
+                /* there should only be one fullscreen session */
+                if (state == null) {
+                    activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_USER
+                    activity.window?.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
+                    return@collect
+                }
+
+                if (store.state.findCustomTabOrSelectedTab(tabId)?.id == state.id) {
+                    setOrientationForTabState(state)
+                }
+                setDeviceSleepModeForTabState(state)
```

Comment:
This whole block has a lot more clarity to me now, thanks for indulging me!

Reply:
Thank you for noticing that it wasn't that clear and helping me improve it!